File: typescript/packages/management/src/tools/add_user_mcp_source.ts
import type { ToolContext, ToolHandler } from '../registrar.js';

type RegisterUserMcpRequest = Parameters<
  ToolContext['client']['sources']['registerUserMcp']
>[1];

export const addUserMcpSource: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const authKind = String(input.auth_kind) as RegisterUserMcpRequest['auth_kind'];
  const request: RegisterUserMcpRequest = {
    name: String(input.name),
    base_url: String(input.base_url),
    auth_kind: authKind,
  };
  if (typeof input.token === 'string') {
    request.token = input.token;
  }
  if (typeof input.auth_header === 'string') {
    request.auth_header = input.auth_header;
  }
  if (typeof input.app_namespace === 'string') {
    request.app_namespace = input.app_namespace;
  }
  if (Array.isArray(input.allowed_tools)) {
    request.allowed_tools = input.allowed_tools.map(String);
  }
  if (Array.isArray(input.ingest_tool_calls)) {
    request.ingest_tool_calls = input.ingest_tool_calls as Array<
      Record<string, unknown>
    >;
  }
  if (typeof input.rate_cap_per_minute === 'number') {
    request.rate_cap_per_minute = input.rate_cap_per_minute;
  }
  if (typeof input.webhook_rate_cap_per_minute === 'number') {
    request.webhook_rate_cap_per_minute = input.webhook_rate_cap_per_minute;
  }
  return ctx.client.sources.registerUserMcp(ctx.sandboxId, request);
};

File: python/copass-management/src/copass_management/tools/add_user_mcp_source.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def add_user_mcp_source(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    kwargs: Dict[str, Any] = {
        "name": str(input["name"]),
        "base_url": str(input["base_url"]),
        "auth_kind": str(input["auth_kind"]),
    }
    if input.get("token") is not None:
        kwargs["token"] = str(input["token"])
    if input.get("auth_header") is not None:
        kwargs["auth_header"] = str(input["auth_header"])
    if input.get("app_namespace") is not None:
        kwargs["app_namespace"] = str(input["app_namespace"])
    if isinstance(input.get("allowed_tools"), list):
        kwargs["allowed_tools"] = [str(x) for x in input["allowed_tools"]]
    if isinstance(input.get("ingest_tool_calls"), list):
        kwargs["ingest_tool_calls"] = list(input["ingest_tool_calls"])
    if input.get("rate_cap_per_minute") is not None:
        kwargs["rate_cap_per_minute"] = int(input["rate_cap_per_minute"])
    if input.get("webhook_rate_cap_per_minute") is not None:
        kwargs["webhook_rate_cap_per_minute"] = int(
            input["webhook_rate_cap_per_minute"]
        )
    return await ctx.client.sources.register_user_mcp(ctx.sandbox_id, **kwargs)
